### ISSUE TITLE
Not reproducible

### ISSUE THREAD

**Issue by stephanlachnit** [jrmadsen/PTL#8]
PTL 1.0.0 is currently not reproducible. The diffs are created by the template `cmake/Templates/PTLConfig.cmake.in`:

https://github.com/jrmadsen/PTL/blob/6049e69b6641ab424150b1a1f46930a35936c728/cmake/Templates/PTLConfig.cmake.in#L41-L42
https://github.com/jrmadsen/PTL/blob/6049e69b6641ab424150b1a1f46930a35936c728/cmake/Templates/PTLConfig.cmake.in#L84
https://github.com/jrmadsen/PTL/blob/6049e69b6641ab424150b1a1f46930a35936c728/cmake/Templates/PTLConfig.cmake.in#L90

The problem is that the current build path is always inserted into `PTLConfig.cmake`, even if it is a system install which shouldn't be the case. I'm not fluent enough in CMake to propose a solution though.

**Comment by jrmadsen** [comment]
Could you clarify what you mean by "not reproducible"? 

The cmake code on line 84 and 90 _shouldn't_ ever be used after in the installation. That second block is basically only used when building PTL itself. If a CMake package is doing `find_package(PTL)`, finding the installed version of PTL, and using those lines, that's definitely a bug. 

Could you show me the error you are getting? No need to try and find the fix yourself, I've been writing CMake for years.

**Comment by stephanlachnit** [comment]
> Could you clarify what you mean by "not reproducible"?

In Debian we test the reproducibility of packages by building them twice with small variations of the system, for example different build times and build folders. For PTL `/usr/lib/x86_64-linux-gnu/PTL/PTLConfig.cmake` differs between these two builds.

In this case being not reproducible isn't that big of a security concern as with other reproducibility issues (see [here](https://reproducible-builds.org/)), but it still triggers the detection which makes finding troublesome reproducibility diffs harder.

> The cmake code on line 84 and 90 _shouldn't_ ever be used after in the installation. That second block is basically only used when building PTL itself. If a CMake package is doing `find_package(PTL)`, finding the installed version of PTL, and using those lines, that's definitely a bug.
>
> Could you show me the error you are getting? No need to try and find the fix yourself, I've been writing CMake for years.

Thanks for offering the help. Maybe I was using the wrong build options, I'll give you more detailed info (sorry for not doing that before). I'm using CMake 3.18.4 and these build options:

`cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_RUNSTATEDIR=/run -DCMAKE_SKIP_INSTALL_ALL_DEPENDENCY=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu ..`

They are so long because these are the default Debian build options for CMake, I can easily overwrite them.

The build path (`/home/stephan/Projects/debian/ptl/ptl`) is in the `else()` part of `if(_PTL_INSTALL_TREE)`. As far as I see it that part is never used, but being in there makes it not producible. Here's the resulting `PTLConfig.cmake` file:

```cmake
#------------------------------------------------------------------------------#
#
#
cmake_policy(SET CMP0012 NEW)

get_filename_component(PTL_CMAKE_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)

include(${CMAKE_CURRENT_LIST_DIR}/PTLConfigVersion.cmake)
include(CMakeFindDependencyMacro)

set(_PTL_AVAILABLE_COMPONENTS )

if(ON)
    list(APPEND _PTL_AVAILABLE_COMPONENTS shared)
endif()

if(ON)
    list(APPEND _PTL_AVAILABLE_COMPONENTS static)
endif()

#------------------------------------------------------------------------------#
#   package initialization
#

####### Expanded from @PACKAGE_INIT@ by configure_package_config_file() #######
####### Any changes to this file will be overwritten by the next CMake run ####
####### The input file was PTLConfig.cmake.in                            ########

get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../" ABSOLUTE)

# Use original install prefix when loaded through a "/usr move"
# cross-prefix symbolic link such as /lib -> /usr/lib.
get_filename_component(_realCurr "${CMAKE_CURRENT_LIST_DIR}" REALPATH)
get_filename_component(_realOrig "/usr/lib/x86_64-linux-gnu/PTL" REALPATH)
if(_realCurr STREQUAL _realOrig)
  set(PACKAGE_PREFIX_DIR "/usr")
endif()
unset(_realOrig)
unset(_realCurr)

macro(set_and_check _var _file)
  set(${_var} "${_file}")
  if(NOT EXISTS "${_file}")
    message(FATAL_ERROR "File or directory ${_file} referenced by variable ${_var} does not exist !")
  endif()
endmacro()

macro(check_required_components _NAME)
  foreach(comp ${${_NAME}_FIND_COMPONENTS})
    if(NOT ${_NAME}_${comp}_FOUND)
      if(${_NAME}_FIND_REQUIRED_${comp})
        set(${_NAME}_FOUND FALSE)
      endif()
    endif()
  endforeach()
endmacro()

####################################################################################


#------------------------------------------------------------------------------#
#   install or build tree
#
set(_PTL_BUILD_TREE )
if(NOT _PTL_BUILD_TREE)
    set(_PTL_INSTALL_TREE ON)
endif()


#------------------------------------------------------------------------------#
#   configure lib and include
#
if(_PTL_INSTALL_TREE)

    set_and_check(PTL_INCLUDE_DIR "${PACKAGE_PREFIX_DIR}/include")
    set_and_check(PTL_LIB_DIR "${PACKAGE_PREFIX_DIR}/lib/x86_64-linux-gnu")

    include(${PTL_CMAKE_DIR}/PTLTargets.cmake)
    check_required_components(PTL)

    # get target
    foreach(_COMPONENT ${_PTL_AVAILABLE_COMPONENTS})
        # set the library
        if(NOT PTL_LIBRARY)
            set(PTL_LIBRARY PTL::ptl-${_COMPONENT} CACHE INTERNAL
                "PTL library")
        endif()
    endforeach()

    if(TRUE)
        set(CMAKE_THREAD_PREFER_PTHREAD ON)
        set(THREADS_PREFER_PTHREAD_FLAG ON)
        find_package(Threads REQUIRED)
    endif()

    if(TRUE)
        set(TBB_ROOT_DIR /usr)
        set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/Modules ${CMAKE_MODULE_PATH})
        find_package(TBB 2020.3 REQUIRED)
    endif()

else()

    set(PTL_INCLUDE_DIR "/home/stephan/Projects/debian/ptl/ptl/source")
    set(PTL_LIB_DIR "/home/stephan/Projects/debian/ptl/ptl/obj-x86_64-linux-gnu")

    set(PTL_Threads_FOUND TRUE)
    set(PTL_TBB_FOUND TRUE)

    if(NOT Threads_FOUND AND PTL_Threads_FOUND)
        set(CMAKE_THREAD_PREFER_PTHREAD ON)
        set(THREADS_PREFER_PTHREAD_FLAG ON)
        find_package(Threads REQUIRED)
    endif()

    if(NOT TBB_FOUND AND PTL_TBB_FOUND)
        set(TBB_ROOT_DIR /usr)
        set(CMAKE_MODULE_PATH /home/stephan/Projects/debian/ptl/ptl/cmake/Modules ${CMAKE_MODULE_PATH})
        find_package(TBB 2020.3 REQUIRED)
    endif()

    foreach(_COMPONENT ${_PTL_AVAILABLE_COMPONENTS})
        if(NOT TARGET ptl-${_COMPONENT})
            include("/home/stephan/Projects/debian/ptl/ptl/obj-x86_64-linux-gnu/PTL-${_COMPONENT}.cmake")
        endif()
        # set the library
        if(NOT PTL_LIBRARY)
            set(PTL_LIBRARY ptl-${_COMPONENT})
        endif()
    endforeach()

    set(PROJECT_DEPENDS ${PTL_LIBRARY})

endif()


#------------------------------------------------------------------------------#
# internal PTL include directory
#
set(PTL_INCLUDE_DIRS ${PTL_INCLUDE_DIR})


#------------------------------------------------------------------------------#
#   CXX standard
#
if(DEFINED CMAKE_CXX_STANDARD)
    if("${CMAKE_CXX_STANDARD}" VERSION_LESS 11)
        set(CMAKE_CXX_STANDARD 11)
        set(CMAKE_CXX_STANDARD_REQUIRED ON)
    endif()
else(DEFINED CMAKE_CXX_STANDARD)
    set(CMAKE_CXX_STANDARD 11)
endif(DEFINED CMAKE_CXX_STANDARD)

set(CMAKE_CXX_STANDARD_REQUIRED ON)


#------------------------------------------------------------------------------#
#   Linking type
#
if(WIN32)
    add_definitions(-D_PTL_ARCHIVE)
endif()


#------------------------------------------------------------------------------#
#   link target
#
set(PTL_LIBRARIES ${PTL_LIBRARY})

#------------------------------------------------------------------------------#
#   Definitions
#
set(PTL_USE_TBB ON)
set(PTL_USE_GPU OFF)
set(PTL_USE_CUDA OFF)

if(NOT _PTL_INSTALL_TREE)
    list(APPEND PTL_LIBRARIES "")
endif()

# message(STATUS "PTL_LIBRARIES: ${PTL_LIBRARIES}")
if(NOT TARGET PTL::ptl)
    add_library(PTL::ptl INTERFACE IMPORTED)
    target_include_directories(PTL::ptl INTERFACE ${PTL_INCLUDE_DIRS})
    target_link_libraries(PTL::ptl INTERFACE ${PTL_LIBRARIES})
    set_target_properties(PTL::ptl PROPERTIES
        INTERFACE_CXX_STANDARD 11
        INTERFACE_CXX_STANDARD_REQUIRED ON)
    if(WIN32)
        target_compile_definitions(PTL::ptl INTERFACE _PTL_ARCHIVE)
    endif()
    foreach(_OPT PTL_USE_TBB PTL_USE_GPU PTL_USE_CUDA)
        if(${_OPT})
            target_compile_definitions(PTL::ptl INTERFACE ${_OPT})
        endif()
    endforeach()
endif()

#------------------------------------------------------------------------------#
#   find_package variables
#
include(FindPackageHandleStandardArgs)
# handle the QUIETLY and REQUIRED arguments and set UnitTest++_FOUND to TRUE
# if all listed variables are TRUE
find_package_handle_standard_args(PTL DEFAULT_MSG
    PTL_CMAKE_DIR
    PTL_LIBRARIES
    PTL_INCLUDE_DIRS)

```

**Comment by jrmadsen** [comment]
> In Debian we test the reproducibility of packages by building them twice with small variations of the system, for example different build times and build folders

Ah, yeah, when you said "reproducible", my background caused me to think you were referring to numerical reproducibility. 

This is pretty easy to accommodate, I'll just generate one `PTLConfig.cmake` for the build tree and one for the install tree.

**Comment by jrmadsen** [comment]
Released v1.0.1

**Comment by stephanlachnit** [comment]
Thanks!
